backend/services/implementations/testSessionService.ts

```diff
@@ -334,10 +334,41 @@ class TestSessionService implements ITestSessionService {
   async updateTestSession(
     id: string,
     testSession: TestSessionRequestDTO,
+    nowDate?: Date,
   ): Promise<TestSessionResponseDTO> {
     let updatedTestSession: TestSession | null;
 
     try {
+      const oldTestSession: TestSession | null = await MgTestSession.findById(
+        id,
+      );
+      if (!oldTestSession) {
+        throw new Error(`Test Session id ${id} not found`);
+      }
+
+      const date = nowDate ?? new Date();
+
+      // If the test session is past, we don't allow teachers to update the test session at all
+      if (oldTestSession.endDate < date) {
+        throw new Error(
+          `Test Session id ${id} has already ended and so cannot be updated`,
+        );
+      }
+
+      // If the test session is active, we don't allow teachers to update the assessment, class, or start date
+      const isActive = oldTestSession.startDate <= date;
+      if (
+        isActive &&
+        (oldTestSession.test.toString() !== testSession.test ||
```

Comment:
Manually enumerating the fields here seems a bit error prone in case we had to add more fields. Maybe convert the object to JSON and use `Object.entries` to enumerate? Something like:

```
const isInvalidModification = Object.entries.some(...)
```
With exclusions in the .some for the keys we allow